docs/config.md
## Configuration

This is content of an example default configuration file. You can also find it [fixtures/config.json](fixtures/config.json)

```json
{
	"redirect_separator": "",
	"authorization_expire": 200,
	"access_expire": 200,
	"allow_get_access": false,
	"allowed_access_type": [
		"authorization_code",
		"refresh_token",
		"password",
		"client_credentials",
		"assertion"
	],
	"token_type": "",
	"provider_name": "",
	"auth_endpoint": "/authorize",
	"token_endpoint": "/tokens",
	"info_endpoint": "/info",
	"port": 8090,
	"database_dialect": "postgres",
	"database_connection": "postgres://postgres:postgres@localhost/hero_test?sslmode=disable",
	"templates_dir": "views",
	"static_dir": "",
	"session_path": "/",
	"session_max_age": 0,
	"session_domain": "",
	"session_secure": false,
	"session_hhhponly": false,
	"session_name": "_hero",
	"Login_template": "login.html",
	"error_template": "error.html",
	"register_template": "",
	"client_template": "client.html",
	"profile_template": "profile.html",
	"home_template": "home.html"
}
```

Table to explain the configuration settings

setting               | type      | details
----------------------|-----------|----------------------
redirect_separator    |  string   | character used to separate multiple redirect urls e.g `:`
authorization_expire  |  int64    | duration in seconds of the authorization code
access_expire         |  int64    | duration in seconds of the access code
allow_get_access      |  bool     | if true allow GET requests
AllowedAccess_type    |  []string | allowed access types e.g `["refresh_token","password"]`
token_type            |  string   | the type of tokens
provider_name         |  string   | the name of the provider( your server) e.g hero
auth_endpoint         |  string   | the route where authorization is handeld e.g /auhorize
token_endpoint        |  string   | the route where access is handled e.g /tokens
info_endpoint         |  string   | the route where the granted user details are accessed.
port                  |  int      | port number where the server will be listenig to. e.g 8080
database_dialect      |  string   | type of database e.g postgres, mysql, foundation
database+connection   |  string   | database connection url
templates_dir         |  string   | the directory where templates are stored.
static_dir            |  string   | the directory where static assets are stored i.e javascript, stylesheets  etc
session_path          |  string   | path of the session( cookie)
session_max_age       |  int      | duration of the cookie(session)
session_domain        |  string   | deomain name for the session
session_secure        |  bool     | true if the sessio is secure( https)
session_httponly      |  bool     | true if session is hhtp only
session_name          |  string   | the name of the session e.g _a_mist_of_avalon
login_template        |  string   | the name of the template to render for login users
error_template        |  string   | the name of the template to render on errors
register_template     |  string   | the name of the template to render on registering users
cLient_template       |  string   | the name of template to render on create/read/update/delete clients
profile_template      |  string   | the name of the template to render on user profile
home_template         |  string   | the name of the template to render at home page



config.go
package hero

import (
	"io/ioutil"
	"path/filepath"
)

//Config conains configuration settings for hero.
type Config struct {
	RedirSeparator      string   `json:"redirect_separator"`
	AuthorizationExpire int64    `json:"authorization_expire"`
	AccessExpire        int64    `json:"access_expire"`
	AllowGetAccess      bool     `json:"allow_get_access"`
	AllowedAccessType   []string `json:"allowed_access_type"`
	TokenType           string   `json:"token_type"`
	ProviderName        string   `json:"provider_name"`
	AuthEndpoint        string   `json:"auth_endpoint"`
	TokenEndpoint       string   `json:"token_endpoint"`
	InfoEndpoint        string   `json:"info_endpoint"`
	Port                int      `json:"port"`
	DatabaseDialect     string   `json:"database_dialect"`
	DatabaseConnection  string   `json:"database_connection"`
	TemplatesDir        string   `json:"templates_dir"`
	StaticDir           string   `json:"static_dir"`
	SessionPath         string   `json:"session_path"`
	SessionMaxAge       int      `json:"session_max_age"`
	SessionDomain       string   `json:"session_domain"`
	SessionSecure       bool     `json:"session_secure"`
	SessionHTTPOnly     bool     `json:"session_hhhponly"`
	SessionName         string   `json:"session_name"`
	LoginTemplate       string   `json:"Login_template"`
	ErrorTemplate       string   `json:"error_template"`
	RegisterTemplate    string   `json:"register_template"`
	CLientTemplate      string   `json:"client_template"`
	ProfileTemplate     string   `json:"profile_template"`
	HomeTemplate        string   `json:"home_template"`
	DocsDir             string   `json:"docs_dir"`
	CsrfSecret          string   `json:"csrf_secret"`
}

// AccessAllowed returns true if accesType is allowed.
func (c *Config) AccessAllowed(acessType string) bool {
	var found bool
	for _, k := range c.AllowedAccessType {
		if k == acessType {
			found = true
			break
		}
	}
	return found
}

// GetDoc reads the content of a file named name found inside the *Config.DocsDir
// directory. This is a convenience
// to be used in templates especially inserting contents from markdown files,
// used in combination with other tempalates functions.
func (c *Config) GetDoc(name string) string {
	if c.DocsDir == "" {
		c.DocsDir = "docs"
	}
	fName := filepath.Join(c.DocsDir, name)
	b, err := ioutil.ReadFile(fName)
	if err != nil {
		// Log this?
		return err.Error()
	}
	return string(b)
}

//DefaultConfig returns *Config with default values.
func DefaultConfig() *Config {
	return &Config{
		AllowedAccessType: []string{
			"authorization_code", "refresh_token",
			"password", "client_credentials",
			"assertion",
		},
		TokenType:           "Bearer",
		AuthorizationExpire: 200,
		AccessExpire:        200,
		AuthEndpoint:        "/authorize",
		TokenEndpoint:       "/tokens",
		InfoEndpoint:        "/info",
		DatabaseConnection:  "postgres://postgres:postgres@localhost/hero_test?sslmode=disable",
		DatabaseDialect:     "postgres",
		ErrorTemplate:       "error.html",
		LoginTemplate:       "login.html",
		RegisterTemplate:    "register.html",
		CLientTemplate:      "client.html",
		ProfileTemplate:     "profile.html",
		HomeTemplate:        "home.html",
		TemplatesDir:        "views",
		SessionPath:         "/",
		SessionName:         "_hero",
		Port:                8090,
		CsrfSecret:          "w4PYxQjVP9ZStjWpBt5t28CEBmRs8NPx",
	}
}
